src/migrate.c

```diff
@@ -0,0 +1,390 @@
+#include <string.h>
+
+#include "redisraft.h"
+
+static bool validSlotMigrationSessionKey(RedisRaftCtx *rr, unsigned int slot, unsigned long long migration_session_key)
+{
+    /* validSlot will have already validated that slot is importing */
+    ShardGroup *sg = rr->sharding_info->importing_slots_map[slot];
+
+    for (unsigned int i = 0; i < sg->slot_ranges_num; i++) {
+        ShardGroupSlotRange *sr = &sg->slot_ranges[i];
+        if (sr->start_slot <= slot && sr->end_slot >= slot) {
+            if (sr->migration_session_key == migration_session_key) {
+                LOG_VERBOSE("migration_session keys matched");
+                return true;
+            } else {
+                LOG_VERBOSE("migration_session keys didn't match, got %llu expected %llu", migration_session_key, sr->migration_session_key);
+                return false;
+            }
+        }
+    }
+
+    LOG_VERBOSE("didn't find shardgroup slot range for this");
+    return false;
+}
+
+static int validSlotTerm(RedisRaftCtx *rr, int slot, raft_term_t term)
+{
+    ShardingInfo *si = rr->sharding_info;
+
+    if (si->max_importing_term[slot] <= term) {
+        si->max_importing_term[slot] = term;
+        return 1;
+    }
+
+    return 0;
+}
+
+static int validSlot(RedisRaftCtx *rr, int slot)
+{
+    ShardGroup *sg = rr->sharding_info->importing_slots_map[slot];
+
+    if (sg && sg->local) {
+        return 1;
+    }
+
+    return 0;
+}
+
+void importKeys(RedisRaftCtx *rr, raft_entry_t *entry)
+{
+    RedisModule_Assert(entry->type == RAFT_LOGTYPE_IMPORT_KEYS);
+
+    RaftReq *req = entry->user_data;
+
+    ImportKeys import_keys = {0};
+    RedisModule_Assert(RaftRedisDeserializeImport(&import_keys, entry->data, entry->data_len) == RR_OK);
+    RedisModule_Assert(import_keys.num_keys > 0);
+
+    // FIXME: validate no cross slot migration at append time
```

Comment:
Is this comment still relevant?

Reply:
removed all sg code

Reply:
https://github.com/RedisLabs/redisraft/issues/411